Address Safer cpp failures in ApplePaySession

iidmsa opened the pull request:
#### 095d7c6f2ef9327475885408872422df862a7946
<pre>
Address Safer cpp failures in ApplePaySession
<a href="https://bugs.webkit.org/show_bug.cgi?id=290127">https://bugs.webkit.org/show_bug.cgi?id=290127</a>
<a href="https://rdar.apple.com/problem/147528417">rdar://problem/147528417</a>

Reviewed by Chris Dumez.

Fix a clang static analyzer warning in ApplePaySession.

* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::ApplePaySession::begin):
(WebCore::ApplePaySession::abort):
(WebCore::ApplePaySession::completeMerchantValidation):
(WebCore::ApplePaySession::completeShippingMethodSelection):
(WebCore::ApplePaySession::completeShippingContactSelection):
(WebCore::ApplePaySession::completePaymentMethodSelection):
(WebCore::ApplePaySession::completeCouponCodeChange):
(WebCore::ApplePaySession::completePayment):
(WebCore::ApplePaySession::didSelectShippingMethod):
(WebCore::ApplePaySession::didSelectShippingContact):
(WebCore::ApplePaySession::didSelectPaymentMethod):
(WebCore::ApplePaySession::didChangeCouponCode):
(WebCore::ApplePaySession::stop):
(WebCore::ApplePaySession::suspend):
(WebCore::ApplePaySession::paymentCoordinator const):
* Source/WebCore/Modules/applepay/ApplePaySession.h:

Canonical link: <a href="https://commits.webkit.org/292497@main">https://commits.webkit.org/292497@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f2c478941c5399ad8f6c8e593d75607dee63f91

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15733 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101183 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46628 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98164 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16028 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24166 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73271 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30488 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99122 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12008 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86822 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53606 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11758 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4578 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45963 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81892 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4680 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103209 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23186 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16883 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82309 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23437 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82839 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81683 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20526 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26299 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3728 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16542 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23149 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28304 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22808 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26288 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24549 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/applepay/ApplePaySession.h:
```diff
@@ -126,7 +126,7 @@ class ApplePaySession final : public PaymentSession, public ActiveDOMObject, pub
     void didCancelPaymentSession(PaymentSessionError&&) override;
 
-    PaymentCoordinator& paymentCoordinator() const;
+    Ref<PaymentCoordinator> paymentCoordinator() const;
```
No, this should keep returning a `PaymentCoordinator&`

cdumez replied:
We should add a new function named `protectedPaymentCoordinator()` which returns a Ref. This is how we usually do things.

cdumez commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -612,10 +612,10 @@ ExceptionOr<void> ApplePaySession::begin(Document& document)
         return Exception { ExceptionCode::InvalidAccessError, "Page already has an active payment session."_s };
 
-    if (!paymentCoordinator().beginPaymentSession(document, *this, m_paymentRequest))
+    if (!protectedPaymentCoordinator()->beginPaymentSession(document, *this, m_paymentRequest))
```
We should cache it in a local variable instead of calling this function twice. This causes unnecessary ref-counting churn otherwise.

cdumez commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -654,11 +654,11 @@ ExceptionOr<void> ApplePaySession::completeMerchantValidation(JSC::JSGlobalObjec
     // PaymentMerchantSession::fromJS() may run JS, which may abort the request so we need to
     // make sure we still have an active session.
-    if (!paymentCoordinator().hasActiveSession())
+    if (!protectedPaymentCoordinator()->hasActiveSession())
```
Ditto.

cdumez commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -1001,6 +1001,11 @@ PaymentCoordinator& ApplePaySession::paymentCoordinator() const
 
+Ref<PaymentCoordinator> ApplePaySession::protectedPaymentCoordinator() const
+{
+    return downcast<Document>(*scriptExecutionContext()).page()->paymentCoordinator();
```
You can just call `return paymentCoordinator();` here.

cdumez reviewed (approved):
LGTM

iidmsa committed:
Address Safer cpp failures in ApplePaySession
https://bugs.webkit.org/show_bug.cgi?id=290127
rdar://problem/147528417

Reviewed by Chris Dumez.

Fix a clang static analyzer warning in ApplePaySession.

* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::ApplePaySession::begin):
(WebCore::ApplePaySession::abort):
(WebCore::ApplePaySession::completeMerchantValidation):
(WebCore::ApplePaySession::completeShippingMethodSelection):
(WebCore::ApplePaySession::completeShippingContactSelection):
(WebCore::ApplePaySession::completePaymentMethodSelection):
(WebCore::ApplePaySession::completeCouponCodeChange):
(WebCore::ApplePaySession::completePayment):
(WebCore::ApplePaySession::didSelectShippingMethod):
(WebCore::ApplePaySession::didSelectShippingContact):
(WebCore::ApplePaySession::didSelectPaymentMethod):
(WebCore::ApplePaySession::didChangeCouponCode):
(WebCore::ApplePaySession::stop):
(WebCore::ApplePaySession::suspend):
(WebCore::ApplePaySession::paymentCoordinator const):
* Source/WebCore/Modules/applepay/ApplePaySession.h:

Canonical link: https://commits.webkit.org/292497@main
